<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd" >
<html>
<head>
<title>EdgeExpressDB - About</title>
    <link href="/4/css/f4_style.css" rel="stylesheet" type="text/css">
    <link href="/4/edgeexpress/css/edgeXpress.css" rel="stylesheet" type="text/css">
</head>
<body>
<div id="mainContainer">

<!-- START OF HEADER -->
<div id="header">
<div id="logo-left">
<img src="/4/images/fantom4-logo.png" style="float:left;">
</div>
<div id="logo-right">
<a href="http://www.osc.riken.jp"><img src="/4/images/osc_color.png" style="float:right;"></a> <a href="http://www.riken.jp"><img src="/4/images/riken-logo.png" style="float:right;"></a>
</div>

<div id="top_menu" style="clear:both;"> 
    <a href="/4/index.php">Home</a>
    <a href="/4/gev/gbrowse/hg18">Genome Browser</a>
    <a class="selected" href="/4/edgeexpress/view">EdgeExpressDB</a>
    <a href="/4/download/">Data</a>
    <a href="/4/paper.php">Paper</a>
</div>

<!-- Sub Header for EdgeExpressDB -->
<div id="exdb_menu">
<div class="text">
<span class="geneStyle">FANTOM4 EdgeExpressDB</span>
<a class="selected" href="../about">about</a>
<a href="../view/#5558263">center view</a>
<a href="../subnet">subnet view</a>
<a href="../contents">contents</a>
</div>
</div>
<!-- End Sub Header -->

<!--               -->
</div>
<!-- END OF HEADER -->


<!-- START OF MAIN CONTENTS OF THE WEB PAGE -->
<div id="mainContents">
<!--                                        -->
<div style="margin: 0% 5% 0% 5%; font-size:14px"> 
<h3>Welcome to EdgeExpressDB</h3>

<p>EdgeExpressDB is a novel database and set of interfaces for systems biology designed for interpreting and comparing biological networks and large high-throughput expression datasets that requires minimal development for novel data types or new search patterns. 

<p>It was developed as part of the FANTOM4 project (Suzuki et al. 2009 â€“ Nature Genetics, submitted).  For FANTOM4 the temporal expression patterns of mRNAs, microRNAs, and promoters, were measured in a human myeloid leukemia cell line undergoing differentiation. Using the expression profiles, and promoter regions identified by deep sequencing of Cap Analysis of gene expression (deepCAGE), expression weighted transcription factor binding sites were predicted. Thereby allowing us to predict regulatory EDGES between a transcription factor and a target promoter (of a gene/miRNA), and place a weight on the likelihood that the transcription factor truly regulates the expression of the promoter.

<p>In addition to predicted TF-> target gene edges we also assembled a combination of siRNA perturbation edges, chromatin immunoprecipitation edges, protein-protein interaction edges and miRNA target gene edges. 
For an introduction to the edge types and weights used in EdgeExpressDB, please refer to <a href="./Supplementary_note_1.pdf">supplementary note 1 (pdf link)</a>, and for an example of how to use interpret the gene centric and subnetwork views please refer to <a href="./Supplementary_note_2.pdf">supplementary note 2(pdf link)</a> which shows the example of CD14,a key monocytic marker.

<p>Contacts:
<ul>
<li>Jessica Severin severin(at)gsc.riken.jp</li>
<li>Andrew Waterhouse andrew(at)gsc.riken.jp</li>
<li>Alistair Forrest forrest(at)gsc.riken.jp</li>
</ul>

</div>

<!--                                     -->
</div>
<!--END OF MAIN CONTENTS OF THE WEB PAGE -->

<!-- START OF FOOTER -->
<div id="footer">
<!--                 -->
  <table width="95%" border="0" cellspacing="0" cellpadding="0">
  <tr>
    <td nowrap><div align="left"><a href="http://www.osc.riken.jp">RIKEN Omics Sciences Center (OSC) RIKEN Yokohama Institute</a>&nbsp;</div></td>
    <td nowrap>Copyright &copy; 2008 RIKEN, Japan. All rights reserved.&nbsp;</td>
    <td nowrap><div align="right"><a href="#header">Back to the top<img src="/4/images/up.gif" align="absmiddle"/></a></div></td>
  </tr>
</table>
<!--               -->
</div>
<!-- END OF FOOTER -->
</div>

</body>
</html>

